### PR TITLE
feat: composable logits processor

### PR DESCRIPTION
**Pull request by xslingcn**
Implement composable logits processor as proposed [here](https://www.notion.so/Logits-Processor-Pipeline-Compiler-Design-Proposal-1fe7b69f364a808b800be5b584818744?pvs=4).

### CODE REVIEW

File: flashinfer/sampling.py [hunk 1]
```diff
@@ -1218,9 +1316,11 @@ def top_k_mask_logits(
     --------
     top_k_renorm_probs
     """
-    return get_sampling_module().top_k_mask_logits(
-        logits, *_to_tensor_scalar_tuple(top_k)
-    )
+    maybe_top_k_arr, top_k_val = _to_tensor_scalar_tuple(top_k)
+    from flashinfer.logits_processor import LogitsPipe, MaskLogits, Softmax, TopK
+
+    pipe = LogitsPipe([Softmax(), TopK(), MaskLogits()])
```
Comment (yzh119) [inline flashinfer/sampling.py]: No need to apply Softmax here.

File: flashinfer/sampling.py [hunk 2]
```diff
@@ -1013,26 +1086,46 @@ def top_k_top_p_sampling_from_probs(
...
     elif filter_apply_order == "joint":
         if check_nan:
             if torch.any(torch.isnan(probs)):
                 raise ValueError("Input probs contains NaN.")
-        return get_sampling_module().top_k_top_p_sampling_from_probs(
+
+        from flashinfer.logits_processor import LogitsPipe, Sampling, TopK, TopP
+
+        pipe = LogitsPipe(
+            [TopK(joint_topk_topp=True), TopP(), Sampling(deterministic=deterministic)]
```
Comment (yzh119) [inline flashinfer/sampling.py]: For joint TopKTopP, it's preferred to have a standalone `Processor`, otherwise we need a argument `joint_topk_topp=` for TopP() as well.

File: flashinfer/logits_processor/fusion_rules.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    pattern: Tuple[type, ...]
+    guard: Callable[[List[Op]], bool]
+    build: Callable[[List[Op]], Op]
+    prio: int = 0
+
+
+def softmax_sampling_guard(window: List[Op]) -> bool:
+    """
+    Logits → Indices
```
Comment (yzh119) [inline flashinfer/logits_processor/fusion_rules.py]:
```suggestion
    Logits -> Indices
```

Other symbol `→ ` should be replaced as well.

File: flashinfer/logits_processor/logits_pipe.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+from .compiler import Compiler, compile_pipeline
+from .fusion_rules import FusionRule
+from .op import Op
+from .types import Sort, TaggedTensor
+from .validators import CompileError, ValidityCheck
+
+
+class LogitsPipe:
+    def __init__(self, ops: List[Op], compiler: Optional[Compiler] = None):
```
Comment (yzh119) [inline flashinfer/logits_processor/logits_pipe.py]: add a standalone function `def compile(self):`, call it when `compile=True`

**Review by yzh119 — COMMENTED** [review]
Add another set of tests called `test_logits_processor` to compare:
1. compare LogitsPipe(compile=True) and LogitsPipe(compile=False)
2. compare sampling ops and its LogitsPipe implementation.

Also, let's separate high level logits processor and Op:
For Op, we should have explicit type `LogitsTopK` and `ProbsTopK` are two different Ops, Op can only have a single input type and a single output type,
We need a `legalize` stage at the begining of compilation to transform high-level logits processor to Ops:
```
# before legalize
Logits -> [TopK(), Sampling()]
# after legalize
Logits -> [LogitsTopKOp(), LogitsSamplingOp()]
```

File: flashinfer/logits_processor/__init__.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+from .op import ParameterizedOp as ParameterizedOp
+from .pipeline import LogitsPipe as LogitsPipe
+from .processors import LogitsProcessor as LogitsProcessor
+from .processors import MinP as MinP
+from .processors import Sample as Sample
+from .processors import Softmax as Softmax
+from .processors import Temperature as Temperature
+from .processors import TopK as TopK
+from .processors import TopP as TopP
+from .types import Sort as Sort
```
Comment (yzh119) [inline flashinfer/logits_processor/__init__.py]: As we discussed yesterday, rename it to TensorType

File: flashinfer/logits_processor/compiler.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        compiled_ops = self._fuse_all(compiled_ops)
+
+        logger.info("Compiled ops: %s", compiled_ops)
+
+        return compiled_ops
+
+    def _type_check(self, ops: List[Op]) -> None:
+        first_op = ops[0]
+
+        current_sort = first_op.IN
```
Comment (yzh119) [inline flashinfer/logits_processor/compiler.py]: current_tensor_type

File: flashinfer/logits_processor/processors.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+limitations under the License.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Any, List
+
+from .types import TensorType
+
+
+class LogitsProcessor(ABC):
```
Comment (yzh119) [inline flashinfer/logits_processor/processors.py:23]: Can you write a docstring with a minimal example?
